pkg/compact/downsample/streamed_block_writer.go

```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "add symbols")
+	}
+
+	// Stream writer is ready to use.
+	w = sw
+	return
+}
+
+// WriteSeries writes chunks data to the chunkWriter, writes lset and chunks Metas to indexWrites and adds label sets to
+// labelsValues sets and memPostings to be written on the finalize state in the end of downsampling process.
+func (w *streamedBlockWriter) WriteSeries(lset labels.Labels, chunks []chunks.Meta) error {
+	if w.finalized || w.ignoreFinalize {
+		return errors.Errorf("series can't be added, writers has been closed or internal error happened")
+	}
+
+	if len(chunks) == 0 {
+		level.Warn(w.logger).Log("empty chunks happened, skip series", lset)
+		return nil
+	}
+
+	if err := w.chunkWriter.WriteChunks(chunks...); err != nil {
+		w.ignoreFinalize = true
+		return errors.Wrap(err, "add chunks")
+	}
+
+	if err := w.indexWriter.AddSeries(w.postings, lset, chunks...); err != nil {
+		w.ignoreFinalize = true
+		return errors.Wrap(err, "add series")
+	}
+
+	w.labelsValues.add(lset)
+	w.memPostings.Add(w.postings, lset)
+	w.postings++
+
+	w.totalChunks += uint64(len(chunks))
+	for i := range chunks {
+		w.totalSamples += uint64(chunks[i].Chunk.NumSamples())
+	}
+
+	return nil
+}
+
+// Close calls finalizer to complete index and meta files and closes all io.CLoser writers.
+// Idempotent.
+func (w *streamedBlockWriter) Close() error {
+	if w.finalized {
+		return nil
+	}
+
+	var merr tsdb.MultiError
+	w.finalized = true
+
+	// Finalise data block only if there wasn't any internal errors.
+	if !w.ignoreFinalize {
+		merr.Add(w.finalize())
+	}
+
+	for _, cl := range w.closers {
+		merr.Add(cl.Close())
+	}
+
+	return errors.Wrap(merr.Err(), "close closers")
+}
+
+// finalize saves prepared index and meta data to corresponding files.
+// It is called on Close. Even if an error happened outside of StreamWriter, it will finalize the block anyway,
+// so it's a caller's responsibility to remove the block's directory.
+func (w *streamedBlockWriter) finalize() error {
+	if err := w.writeLabelSets(); err != nil {
+		return errors.Wrap(err, "write label sets")
+	}
+
+	if err := w.writeMemPostings(); err != nil {
+		return errors.Wrap(err, "write mem postings")
+	}
+
+	if err := w.writeMetaFile(); err != nil {
+		return errors.Wrap(err, "write meta meta")
+	}
+
+	if err := w.syncDir(); err != nil {
+		return errors.Wrap(err, "sync blockDir")
+	}
+
+	level.Info(w.logger).Log(
+		"msg", "write downsampled block",
+		"mint", w.meta.MinTime,
+		"maxt", w.meta.MaxTime,
+		"ulid", w.meta.ULID,
+		"resolution", w.meta.Thanos.Downsample.Resolution,
+	)
+	return nil
+}
+
+// syncDir syncs blockDir on disk.
+func (w *streamedBlockWriter) syncDir() error {
+	df, err := fileutil.OpenDir(w.blockDir)
+	if err != nil {
+		return errors.Wrap(err, "open temporary block blockDir")
+	}
+
+	defer runutil.CloseWithErrCapture(w.logger, &err, df, "close temporary block blockDir")
+
+	err = fileutil.Fsync(df)
```

Comment:
we can do `if err := fileutil.Fsync(df); err != nil` here

Reply:
It will work only with assigning operator `if err = fileutil.Fsync(df); err != nil` as we might need this error in `defer` above, but it's the error prone approach to see the difference between two of those `=` and `:=` 🙃

Reply:
oh this does not work like this! This is wrong. As you might have something in `err` but it will not return an error (: 

We need to `syncDir() (err error)` and then you can just `return nil, whatevererror` and this return arg `err` would have `whatevererror` (: It might be confusing agree.